src/gluonnlp/data/sampler.py

```diff
@@ -266,8 +266,88 @@ def __len__(self):
         return len(self._sorted_ids)
 
 
+class BoundedBudgetSampler(BaseSampler):
+    r"""Assign each data sample to bounded budget batches. Samples will be sorted by length before batchfy
+    see https://github.com/pytorch/fairseq/blob/master/fairseq/data/data_utils_fast.pyx
+
+    Parameters
+    ----------
+    lengths
+        The length of the sequences in the input data sample.
+    max_tokens
+        max tokens num of each batch
+    max_sentences
+        max sentences num of each batch
+    required_batch_size_multiple
+        require batch size to be a multiple of N (default: 1).
+        better throughput in GPU.
+    shuffle
+        Whether to shuffle the batches.
+    seed
+        The seed of the sampler
+    """
+    def __init__(self, lengths: Union[Sequence[int], Sequence[Sequence[int]]],
+                 max_tokens: int = -1, max_sentences: int = -1,
+                 required_batch_size_multiple: int = 1,
+                 shuffle: bool = False, seed: Optional[int] = None):
+        assert len(lengths) > 0, 'BoundedBudgetSampler does not support empty lengths.'
+        assert max_tokens > 0 or max_sentences > 0, 'One of max_tokens and max_sentences must be larger than 0'
+        self._lengths = np.array(lengths)
+        if self._lengths.ndim == 2:
+            self._lengths = self._lengths.max(axis=1)
+        self._indices = np.array(range(len(lengths)))
+        self._max_tokens = max_tokens
+        self._max_sentences = max_sentences
+        self._batches = []
+        self._shuffle = shuffle
+        self._rng = np.random.RandomState(seed)
+        # sort
+        self._indices = self._indices[np.argsort(self._lengths, kind='mergesort')]
+        batch = []
+        # max len in a batch
+        batch_max_sample_len = 0
+        for index in self._indices:
+            batch_max_sample_len = max(batch_max_sample_len, self._lengths[index])
+            # try to insert new sample to the batch
+            batch_num_sentences = len(batch) + 1
+            batch_num_tokens = batch_num_sentences * batch_max_sample_len
+            if (self._max_sentences > 0 and batch_num_sentences > self._max_sentences) or \
+               (self._max_tokens > 0 and batch_num_tokens > self._max_tokens):
+                moded_bs = max(
+                    required_batch_size_multiple * (len(batch) // required_batch_size_multiple),
+                    len(batch) % required_batch_size_multiple
```

Comment:
when this can be larger?

Reply:
this is consistent with
```python
if len(batch) // required_batch_size_multiple == 0:
    moded_bs = len(batch) % required_batch_size_multiple
```

Reply:
Ok. You can add comment there saying moded_bs = len(batch) % required_batch_size_multiple when len(batch) < required_batch_size_multiple